PythonRational optimizations

asmeurer opened the pull request:
These are the optimizations suggested at https://github.com/sympy/sympy/issues/12602. That issue is still incomplete because I would like to also see these things ported to the core (but likely I won't do that in this PR).

asmeurer committed:
Use python_gcd in PythonRational.__init__

asmeurer committed:
Use more efficient "crosswise" gcd algorithm in PythonRational.__mul__ and __div__

See https://github.com/sympy/sympy/issues/12602.

asmeurer committed:
Use more efficient gcd algorithm in PythonRational.__add__ and __sub__

See https://github.com/sympy/sympy/issues/12602. Thanks to @isuruf for the
suggestion.

asmeurer committed:
Use crosswise gcd in PythonRational.__rmul__ and __rdiv__

See https://github.com/sympy/sympy/issues/12602.

asmeurer committed:
Don't recompute the gcd in PythonRational.__radd__ and __rsub__

It only supports integers, and p/q +/- a = (p +/- qa)/q is already in lowest
terms when p/q is in lowest terms.

asmeurer commented on sympy/polys/domains/pythonrational.py:
```diff
@@ -42,7 +42,8 @@ def parent(self):
 
-    def __init__(self, p, q=1):
+    def __init__(self, p, q=1, _gcd=True):
+        from sympy.polys.domains.groundtypes import python_gcd as gcd
```
The imports have to go inside the functions because of circular import issues (groundtypes imports this file).

asmeurer committed:
Merge branch 'master' into python-rational-optimizations

asmeurer commented:
Oh, I didn't actually merge this branch with master correctly, which gives an interesting set of benchmarks (all in Python 3.5):

**master**(remember in master igcd_lehmer and math.gcd isn't integrated with PythonRational at all)
```
In [3]: %time count_roots(expr)
CPU times: user 1min 37s, sys: 398 ms, total: 1min 37s
Wall time: 1min 37s
```

**this branch, before merging master** (the time here is coming from the igcd cache)
```
In [2]: %time count_roots(expr)
CPU times: user 15.9 s, sys: 309 ms, total: 16.2 s
Wall time: 16.5 s
```

**d20bce6 (just the commit adding python_gcd to `PythonRational.__init__`) merged with master**
```
In [2]: %time count_roots(expr)
CPU times: user 5.27 s, sys: 22.9 ms, total: 5.29 s
Wall time: 5.31 s
```

**this branch, after merging master**
```
In [2]: %time count_roots(expr)
CPU times: user 2.08 s, sys: 14.5 ms, total: 2.1 s
Wall time: 2.1 s
```

So the igcd cache resulted in a 5.9x speedup, the math.gcd resulted in a 3.1x speedup, and the crosswise optimizations resulted in a 2.5x speedup. 

Now, the bad news, is that in Python 2, we don't see the math.gcd/igcd_lehmer speedup at all. I guess the numbers are too small to use igcd_lehmer. The times go `1min 37s`, `15.4 s`, `15.8 s`, and `5.24 s`, respectively. 

And of course, it only takes 304 ms when gmpy2 is installed.

asmeurer committed:
Use as_int() in igcd2()

jksuom commented on sympy/core/numbers.py:
```diff
@@ -211,6 +211,7 @@ def igcd(*args):
 except ImportError:
     def igcd2(a, b):
         """Compute gcd of two Python integers a and b."""
+        a, b = as_int(a), as_int(b)
```
I did intentionally not add this line to make `igcd2` work with Python int's only compatibly with `math.gcd`.

asmeurer replied:
Tests were failing without this. I guess the fix should be elsewhere.

smichr replied:
Do you recall the nature of the failing test? As you say, perhaps the recasting should happen in the function that was calling `igcd2`.

jksuom replied:
`igcd2` is not called in many places. I think the obvious suspects are the calls after `if isinstance(other, integer_types)` in `pythonrational`.

asmeurer replied:
It is coming from [here](https://github.com/sympy/sympy/blob/5522219676a18f250582ee39654799d80ef40c24/sympy/polys/ring_series.py#L975). Apparently `PythonRational.__pow__` uses [`self.new`](https://github.com/sympy/sympy/blob/5522219676a18f250582ee39654799d80ef40c24/sympy/polys/domains/pythonrational.py#L201), which is probably the original way of avoiding gcd computation. 

In that code, is `c` always supposed to be an `n`th power (is that what the `DomainError` catches?)? If so, we can replace it with `integer_nthroot`.

jksuom commented:
This looks good. I did try to read carefully the new code lines.

asmeurer committed:
Allow Integer and Rational input to PythonRational

asmeurer committed:
Use __init__(_gcd=False) instead of new in PythonRational.__pow__

The former also works if the input is a SymPy type, which is possible if the
exponent is an Integer or Rational.

asmeurer committed:
Remove as_int checks from igcd2

The caller should do this check. See
https://github.com/sympy/sympy/pull/12656#discussion_r117604184.

asmeurer commented:
OK, I pushed up some fixes. I had to allow Rational and Integer input in PythonRational to make the removal of `as_int` work (the code in ring_series, mentioned in the [above discussion](https://github.com/sympy/sympy/pull/12656#discussion_r117604184), assumes that this works, and I figured this was easier than trying to add cases to ring_series to not do that).

asmeurer committed:
Merge branch 'master' into python-rational-optimizations

smichr commented:
@jksuom said it looked good up to the last commits and the new additions look fine. I agree that the caller should make the args to igcd ints. The docstring also states that it is for Python ints.
